### ISSUE TITLE
no-unmodified-loop-condition too strict?

### ISSUE THREAD

**Issue by mar10** [eslint/eslint#11302]
**What rule do you want to change?**
no-unmodified-loop-condition

**Does this change cause the rule to produce more or fewer warnings?**
Should produce fewer warnings.

**How will the change be implemented? (New option, new default behavior, etc.)?**
new default behavior

**Please provide some example code that this change will affect:**

```js
/* eslint no-unmodified-loop-condition: error */
function foo(opts) {
	var all = opts.all;
	while ( (all || node.expanded) && node.children ) {
		children = node.children;
		parent = node;
		node = children[children.length - 1];
	}
}
```

**What does the rule currently do for this code?**

Shows an error `'all' is not modified in this loop` ([see demo](https://eslint.org/demo/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVubW9kaWZpZWQtbG9vcC1jb25kaXRpb246IGVycm9yICovXHJcbmZ1bmN0aW9uIGZvbyhvcHRzKSB7XHJcblx0dmFyIGFsbCA9IG9wdHMuYWxsO1xyXG5cdHdoaWxlICggKGFsbCB8fCBub2RlLmV4cGFuZGVkKSAmJiBub2RlLmNoaWxkcmVuICkge1xyXG5cdFx0Y2hpbGRyZW4gPSBub2RlLmNoaWxkcmVuO1xyXG5cdFx0cGFyZW50ID0gbm9kZTtcclxuXHRcdG5vZGUgPSBjaGlsZHJlbltjaGlsZHJlbi5sZW5ndGggLSAxXTtcclxuXHR9XHJcbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjksInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=))

**What will the rule do after it's changed?**

IMO there should be no warning,  because the 'loop condition' *is* modified (although not all loop condition *parts* are).
In my understanding, this rule should detect possibly infinite loops. 
I don't see much value in the current interpretation of the rule (maybe you have an example?).

The sample above is a bit stripped down, but should be enough to give an idea of a use case.
#6280 is a related one.

**Are you willing to submit a pull request to implement this change?**

Rather not at the moment, sorry. Some thoughts though:
It might require some analysis of the term.

If we assume that `node` is modified inside the loop, and `all` is unmodified:

This loop-conditions could be considered 'modified':
```js
while ( (all || node.expanded) && node.children ) {
while ( node.expanded ) {
while ( (all && node.expanded) && node.children ) {
```

This loop-condition could be considered 'potentially unmodified' and should be flagged as well:
```js
while ( all || node.expanded ) {. // unmodified if `all` is truthy
while ( all && node.expanded ) { // unmodified if `all` is falsy
```

This loop-condition could be considered 'unmodified':
```js
while ( all ) {
while ( !all ) {
while ( all !== true  ) {
```

**Comment by fineline** [comment]
I have another example of this, where I am using a boolean parameter to change the behavior of a looping function:

```
function getHierarchy(item, traverseSecureItems) {
  const result = []
  while (item && (traverseSecureItems || !item.isSecure)) {
    result.push(item)
    item = item.parent
  }
  return result
}
```
Gives the warning on traverseSecureItems. Only other way I can think to express this is an `if` condition / `break` within the loop - bit ugly.

**Comment by fineline** [comment]
OK, maybe ugly is a bit harsh, it's in the eye of the beholder I guess. 

```
function getHierarchy (item, traverseSecureItems) {
  const result = []
  while (item) {
    if (item.isSecure && !traverseSecureItems) break
    result.push(item)
    item = item.parent
  }
  return result
}
```

I just normally avoid `break`s. 

Still, the rule is warning on a valid use of a non-modified variable in the original version.
